### PR TITLE
feat(dateextract step): Add new date information to be extracted + extract several info in just one step

### PR DESCRIPTION
**Pull request by vdestraitt**
This PR improves the current `dateextract` steps in 2 ways:

- Adding a lot of new convenient date information to be extracted (eg. as first day of a period or previous period). See the detail documented [here](https://github.com/ToucanToco/weaverbird/blob/b63301812629211a064c550acdec26ecb1451fca/docs/_docs/user-interface/dateextract.md)
- Allowing to extract all the required date information in just one step

### CODE REVIEW

File: src/lib/translators/mongo.ts [hunk 1]
```diff
@@ -722,6 +725,204 @@ function transformConcatenate(step: Readonly<S.ConcatenateStep>): MongoStep {
...
+    if (d === 'quarter') {
+      addFields[newColumns[i]] = {
+        $switch: {
+          branches: [
+            { case: { $lte: [{ $divide: [{ $month: $$(step.column) }, 3] }, 1] }, then: 1 },
+            { case: { $lte: [{ $divide: [{ $month: $$(step.column) }, 3] }, 2] }, then: 2 },
+            { case: { $lte: [{ $divide: [{ $month: $$(step.column) }, 3] }, 3] }, then: 3 },
+          ],
+          default: 4,
+        },
```
Comment (austil) [inline src/lib/translators/mongo.ts:754]: Est ce qu'on aurai pas pu utiliser ?

```js
{
  $ceil: { $divide: [{ $month: $$(step.column) }, 3] },
}
```

**Review by austil — COMMENTED** [review]
Pas un petit morceau, comme la query mongo générée est assez massive je vais d'abord aller faire un tour sur le feature peak pour voir le résultat.

File: docs/_docs/user-interface/dateextract.md [hunk 2]
```diff
@@ -1,23 +1,49 @@
...
+- `first day of month`: calendar dat corresponding to the first day of the month,
+- `first day of week`: calendar dat corresponding to the first day of the week,
+- `first day of quarter`: calendar dat corresponding to the first day of the quarter,
+- `first day of ISO week`: calendar dat corresponding to the first day of the week in ISO 8601 format,
+- `previous day`: calendar date one day before the target date,
+- `first day of previous year`: calendar dat corresponding to the first day (1st of January) of the previous year,
+- `first day of previous month`: calendar dat corresponding to the first day of the previous month,
+- `first day of previous week`: calendar dat corresponding to the first day of the previous week,
+- `first day of previous ISO week`: calendar dat corresponding to the first day of the previous ISO week,
+- `first day of previous quarter`: calendar dat corresponding to the first day of the previous quarter,
```
Comment (austil) [inline docs/_docs/user-interface/dateextract.md:37]: Calendar **DAT**

**Review by austil — APPROVED** [review]
C'est massif donc le gros du travail sera plus sur la partie UAT que review de PR.

**Comment by davinov** [comment]
@austil First feedback, the title of the step is not the good one I think
Screenshot from 2021-03-23 10-26-04

File: src/components/stepforms/schemas/dateextract.ts [hunk 3]
```diff
@@ -1,6 +1,4 @@
-import { addNotInColumnNamesConstraint, StepFormType } from './utils';
-
-const schema = {
+export default {
   $schema: 'http://json-schema.org/draft-07/schema#',
   title: 'Extract date propperty from date step',
```
Comment (austil) [inline src/components/stepforms/schemas/dateextract.ts:3]: Need to change this title

**Comment by davinov** [comment]
Thanks fort his awesome work!
I've run some tests and it works great on mongo and pandas backends :)

Only feedback left: when asking for the previous day of... while using the pandas backend, the type of the column return is not date but a string. Do you think we could change that? 
Screenshot from 2021-03-23 11-43-24

**Comment by austil** [comment]
@fspot

**Comment by fspot** [comment]
Hmm d'après le code et les tests c'est bien une date qui est retournée par le previousDay :shrug:

**Comment by austil** [comment]
Pour ma part je n'ai aucune idée de comment le titre du screenshot en question a pû se retrouver là ...

**Review by davinov — APPROVED** [review]
Thanks a lot yo you 3, and congrats @vdestraitt for your last PR :clap:

**Pull request merged**